pkg/controller/master-controller-manager/kubeone/controller.go

```diff
@@ -231,91 +229,220 @@ func (r *reconciler) Reconcile(ctx context.Context, request reconcile.Request) (
 	return r.reconcile(ctx, log, externalCluster)
 }
 
-func (r *reconciler) reconcile(ctx context.Context, log *zap.SugaredLogger, externalCluster *kubermaticv1.ExternalCluster) (reconcile.Result, error) {
+func (r *reconciler) initiateImportAction(ctx context.Context, log *zap.SugaredLogger, externalCluster *kubermaticv1.ExternalCluster) error {
 	kubeconfigSecret := &corev1.Secret{}
 	if err := r.Get(ctx, ctrlruntimeclient.ObjectKey{Namespace: kubernetesprovider.GetKubeOneNamespaceName(externalCluster.Name), Name: resources.KubeOneKubeconfigSecretName}, kubeconfigSecret); err != nil {
 		if kerrors.IsNotFound(err) {
 			if err := r.importCluster(ctx, log, externalCluster); err != nil {
-				return reconcile.Result{}, err
+				return err
 			}
 		}
-		return reconcile.Result{}, err
-	}
-	externalClusterProvider, err := kubernetesprovider.NewExternalClusterProvider(r.ImpersonationClient, r.Client)
-	if err != nil {
-		return reconcile.Result{}, err
+		return err
 	}
+	return nil
+}
+
+func (r *reconciler) upgradeAction(ctx context.Context,
+	log *zap.SugaredLogger,
+	externalCluster *kubermaticv1.ExternalCluster,
+	externalClusterProvider *kubernetesprovider.ExternalClusterProvider,
+	currentManifest []byte,
+	clusterStatus kubermaticv1.Status,
+) (*corev1.Pod, error) {
+	upgradePod := &corev1.Pod{}
 	version, err := externalClusterProvider.GetVersion(ctx, externalCluster)
 	if err != nil {
-		return reconcile.Result{}, err
+		return nil, err
 	}
 	currentVersion := version.String()
+	wantedVersion, err := getWantedVersion(currentManifest)
+	if err != nil {
+		return nil, err
+	}
+	if clusterStatus != kubermaticv1.StatusReconciling && clusterStatus != kubermaticv1.StatusError {
```

Comment:
Nit: If you flip the condition around, you don't have to indent as much code:

```go
if clusterStatus == ... || clusterStatus == ... {
   return nil, nil 
}

// more stuff...
```

Also, even more nitty, it's easier to me to read and comprehend "if thing is A or thing is B" than "if thing is NOT A and thing is NOT B".